_posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

```diff
@@ -0,0 +1,753 @@
+| This determines both how many output features there are *and* the dimensionality of the hidden states.                             |               |              |
++------------------------------------------------------------------------------------------------------------------------------------+---------------+--------------+
+| *Per-time-step output; latent state; intermediate state ...*                                                                       | hidden state  | memory state |
+|                                                                                                                                    |               |              |
+| This could be named "public state" in the sense that we, the users, are able to obtain *all values.*                               |               |              |
++------------------------------------------------------------------------------------------------------------------------------------+---------------+--------------+
+| *Cell state; inner state ... (LSTM only)*                                                                                          | cell state    | carry state  |
+|                                                                                                                                    |               |              |
+| This could be named "private state" in that we are able to obtain a value *only for the last time step*. More on that in a second. |               |              |
++------------------------------------------------------------------------------------------------------------------------------------+---------------+--------------+
+
+: Table 1: RNN terminology. Comparing torch-speak and Keras-speak. In row 1, the terms are parameter names. In rows 2 and 3, they are pulled from current documentation.
+:::
+
+<br />
+
+Now, about that public vs. private distinction. In both frameworks, we can obtain outputs (hidden states) for every time step. The cell state, however, we can access only for the very last time step. This, it need be said, is purely an implementation decision. As we'll see when building our own recurrent module, there are no obstacles inherent in keeping track of cell states and passing them back to the user.
+
+If you dislike the pragmatism of this distinction, you can always go with the math. When a new cell state has been computed (based on prior cell state, input, forget, and cell gates -- the specifics of which we are not going to get into here), it is transformed to the hidden (a.k.a. output) state making use of yet another, the output gate:
+
+$$
+h_t = o_t \odot \tanh(c_t)
+$$
+
+Definitely, then, hidden state (output, resp.) builds on cell state, adding additional modeling power. Whether one would interpret the workings of the output gate in an anthropomorphizing way is left to personal choice.
```

Comment:
This is probably because I don't understand the technical aspects, but I'm not even sure how one would go about interpreting the output gate as though it were human. If you think this last sentence adds something, then you can leave it. But, given the amount of prose in this piece (all prior to actually building the convLSTM), I'd consider cutting it.

Reply:
you're right as always - this is a leftover hinting at an -  even longer - paragraph that got completely removed (because verbosity ... yeah ....)